Update Discovery Protocols documentation [8474] (#139)
eProsima/Fast-RTPS-docs: docs/fastdds/discovery/server_client.rst

.. _discovery_server:

Server-Client Discovery Settings
--------------------------------

This mechanism is based on a client-server discovery paradigm, i.e. the metatraffic (message exchange among
|DomainParticipants| to identify each other) is managed by one or several server |DomainParticipants| (left figure), as
opposed to simple discovery (right figure), where metatraffic is exchanged using a message broadcast mechanism like an
IP multicast protocol.
A `Discovery-Server <https://eprosima-discovery-server.readthedocs.io/en/latest/index.html>`_ tool is available to
ease client-server setup and testing.

.. figure:: /01-figures/fast_dds/discovery/discovery-server.svg
    :align: center
    :width: 50%

    Comparison of Server-Client discovery and Simple discovery mechanisms

.. _DS_key_concepts:

Key concepts
^^^^^^^^^^^^

In this architecture there are several key concepts to understand:

- The Server-Client discovery mechanism reuses the RTPS discovery messages structure, as well as the standard DDS
  |DataWriters| and |DataReaders|.

- Discovery server |DomainParticipants| may be *clients* or *servers*.
  The only difference between them is how they handle meta-traffic.

>   The user traffic, that is, the traffic among the |DataWriters| and |DataReaders| they create is role-independent.

Reviewer: ```suggestion
  The user traffic, that is, the traffic among the |DataWriters| and |DataReaders| they create, is role-independent.
```

  Note that a *server* may act as a *client* for other *servers*.

- *Clients* require a beforehand knowledge of the *servers* to which they want to link.
  Basically it is reduced to the *server* identity (henceforth called ``GuidPrefix``) and a list of locators where the
  *server* is listening.
  These locators also define the transport protocol (UDP or TCP) the client will use to contact the *server*.

  - The ``GuidPrefix`` is the RTPS standard |RTPSParticipant| unique identifier, a 12-byte chain.
    This identifier allows clients to assess whether they are receiving messages from the right server, as each
    standard RTPS message contains this piece of information.

    The ``GuidPrefix`` is used because the server's IP address may not be a reliable enough server identifier,
    since several servers can be hosted in the same machine, thus having the same IP, and also because multicast
    addresses are acceptable addresses.

- *Servers* do not require any beforehand knowledge of their *clients*, but their ``GuidPrefix`` and locator list (where
  they are listening) must match the one provided to the *clients*.

>   In order to gather *client* discovery information the following handshake strategy is followed:

Reviewer: ```suggestion
  In order to gather *client* discovery information, the following handshake strategy is followed:
```

  - *Clients* send discovery messages to the *servers* at regular intervals (ping period) until they receive message
    reception acknowledgement.

  - *Servers* receive discovery messages from the clients, but they do not start processing them until a time interval
    has elapsed, which starts at the moment the server is instantiated.

> In order to clarify this discovery setup, either on compile time (sources) or runtime (XML files), its explanation is
> divided into two sections: focusing on the main concepts (:ref:`setup by concept <DS_setup_concepts>`), and on
> the main setting structures and XML tags (:ref:`setup by QoS <DS_setup_attributes>`).

Reviewer: ```suggestion
In order to clarify this discovery setup, either on compile time (sources) or runtime (XML files), this explanation is
divided into two sections: on focusing on the main concepts (:ref:`setup by concept <DS_setup_concepts>`), and another one focusing on
the main setting structures and XML tags (:ref:`setup by QoS <DS_setup_attributes>`).
```
